File: app/src/main/java/com/tokyonth/installer/install/UnInstallTask.java
package com.tokyonth.installer.install;

import android.content.Context;
import android.os.Handler;

import com.catchingnow.icebox.sdk_client.IceBox;
import com.tokyonth.installer.Constants;
import com.tokyonth.installer.bean.ApkInfoBean;
import com.tokyonth.installer.utils.Shell;
import com.tokyonth.installer.utils.ShellUtils;
import com.tokyonth.installer.utils.ShizukuShell;

public class UnInstallTask extends Thread {

    private String pkgName;
    private int mode;
    private Context context;
    private CommanderCallback commanderCallback;
    private ApkInfoBean apkInfoBean;
    private Handler handler;

    public UnInstallTask(int mode, String pkgName, Context context, Handler handler, CommanderCallback commanderCallback) {
        this.mode = mode;
        this.pkgName = pkgName;
        this.context = context;
        this.handler = handler;
        this.commanderCallback = commanderCallback;
        fakeDate();
    }

    private void fakeDate() {
        apkInfoBean = new ApkInfoBean();
    }

    public void shellMode() {
        int resultCode = ShellUtils.execWithRoot(Constants.INSTANCE.getUNFREEZE_COMMAND() + pkgName);
        commanderCallback.onApkInstalled(apkInfoBean, resultCode);
    }

    @Override
    public void run() {
        super.run();
        switch (mode) {
            case 1:
                ShizukuShell shell = new ShizukuShell();
                Shell.Result result = shell.exec(new Shell.Command("pm", "uninstall", pkgName));
                handler.post(() -> commanderCallback.onApkInstalled(apkInfoBean, result.exitCode));
                break;
            case 2:
                if (IceBox.uninstallPackage(context, pkgName)) {
                    handler.post(() -> commanderCallback.onApkInstalled(apkInfoBean, 0));
                } else {
                    handler.post(() -> commanderCallback.onApkInstalled(apkInfoBean, 1));
                }
                break;
        }
    }

}

File: app/src/main/java/com/tokyonth/installer/install/UninstallTask.kt
package com.tokyonth.installer.install

import android.os.Handler
import android.os.Looper
import com.catchingnow.icebox.sdk_client.IceBox
import com.tokyonth.installer.App
import com.tokyonth.installer.Constants
import com.tokyonth.installer.data.LocalDataRepo
import com.tokyonth.installer.utils.ShellUtils

class UninstallTask(
    private val pkgName: String,
    private val uninstallCallback: UninstallCallback
) : Thread() {

    private val handler: Handler = Handler(Looper.getMainLooper())

    override fun run() {
        super.run()
        when (LocalDataRepo.instance.getInstallMode()) {
            0 -> {
                ShellUtils.execWithRoot(Constants.UNINSTALL_COMMAND + pkgName).let {
                    handler.post { uninstallCallback.onUninstallResult(it) }
                }
            }
            1 -> {
                ShellUtils.execWithRoot(Constants.UNINSTALL_COMMAND + pkgName).let {
                    handler.post { uninstallCallback.onUninstallResult(it) }
                }
            }
            2 -> {
                IceBox.uninstallPackage(App.context, pkgName).let {
                    val result = if (it) {
                        0
                    } else {
                        -1
                    }
                    handler.post { uninstallCallback.onUninstallResult(result) }
                }
            }
        }
    }

}
